service/src/prometheus_metrics.rs

```diff
@@ -0,0 +1,161 @@
+		register_int_gauge!("enclave_account_free_balance", "Free balance of the enclave account")
+			.unwrap();
+	static ref ENCLAVE_SIDECHAIN_BLOCK_HEIGHT: IntGauge =
+		register_int_gauge!("enclave_sidechain_block_height", "Enclave sidechain block height")
+			.unwrap();
+	static ref ENCLAVE_SIDECHAIN_TOP_POOL_SIZE: IntGauge =
+		register_int_gauge!("enclave_sidechain_top_pool_size", "Enclave sidechain top pool size")
+			.unwrap();
+}
+
+pub async fn start_metrics_server<MetricsHandler>(
+	metrics_handler: Arc<MetricsHandler>,
+	port: u16,
+) -> ServiceResult<()>
+where
+	MetricsHandler: HandleMetrics + Send + Sync + 'static,
+{
+	let metrics_route = warp::path!("metrics").and_then(move || {
+		let handler_clone = metrics_handler.clone();
+		async move { handler_clone.handle_metrics().await }
+	});
+	let socket_addr: SocketAddr = ([0, 0, 0, 0], port).into();
+
+	info!("Running prometheus metrics server on: {:?}", socket_addr);
+	warp::serve(metrics_route).run(socket_addr).await;
+
+	info!("Prometheus metrics server shut down");
+	Ok(())
+}
+
+#[async_trait]
+pub trait HandleMetrics {
+	type ReplyType: Reply;
+
+	async fn handle_metrics(&self) -> Result<Self::ReplyType, Rejection>;
+}
+
+/// Metrics handler implementation.
+pub struct MetricsHandler<Wallet> {
+	enclave_wallet: Arc<Wallet>,
+}
+
+#[async_trait]
+impl<Wallet> HandleMetrics for MetricsHandler<Wallet>
+where
+	Wallet: EnclaveWalletInfo + Send + Sync,
+{
+	type ReplyType = String;
+
+	async fn handle_metrics(&self) -> Result<Self::ReplyType, Rejection> {
+		self.update_metrics().await;
+
+		let default_metrics = match gather_metrics_into_reply(&prometheus::gather()) {
+			Ok(r) => r,
+			Err(e) => {
+				error!("Failed to gather prometheus metrics: {:?}", e);
+				String::default()
+			},
+		};
+
+		Ok(default_metrics)
+	}
+}
+
+impl<Wallet> MetricsHandler<Wallet>
+where
+	Wallet: EnclaveWalletInfo + Send + Sync,
+{
+	pub fn new(enclave_wallet: Arc<Wallet>) -> Self {
+		MetricsHandler { enclave_wallet }
+	}
+
+	async fn update_metrics(&self) {
+		match self.enclave_wallet.free_balance() {
+			Ok(b) => {
+				ENCLAVE_ACCOUNT_FREE_BALANCE.set(b as i64);
+			},
+			Err(e) => {
+				error!("Failed to fetch free balance metric, value will not be updated: {:?}", e);
+			},
+		}
+	}
+}
+
+fn gather_metrics_into_reply(metrics: &[MetricFamily]) -> ServiceResult<String> {
+	use prometheus::Encoder;
+	let encoder = prometheus::TextEncoder::new();
+
+	let mut buffer = Vec::new();
+	encoder.encode(metrics, &mut buffer).map_err(|e| {
+		Error::Custom(format!("Failed to encode prometheus metrics: {:?}", e).into())
+	})?;
+
+	let result_string = String::from_utf8(buffer).map_err(|e| {
+		Error::Custom(
+			format!("Failed to convert Prometheus encoded metrics to UTF8: {:?}", e).into(),
+		)
+	})?;
+
+	Ok(result_string)
+}
+
+/// Trait to receive metric updates from inside the enclave.
+pub trait ReceiveEnclaveMetric {
+	fn receive_enclave_metric(&self, metric: EnclaveMetric) -> ServiceResult<()>;
+}
```

Comment:
This is the receiver for enclave metrics (metrics that have been passed through the o-call interface)

Reply:
This all looks so nice and clean